lib/manageiq_performance/commands/profile.rb

```diff
@@ -0,0 +1,268 @@
+
+      DEFAULT_OPTS = {
+        :target_type => :file,
+        :full_require_tree => true,
+        :stackprof_mode => :cpu,
+        :stackprof_interval => 1000
+      }
+
+      def self.help_text
+        "memory/timing profile of a single ruby script or application"
+      end
+
+      def self.run(args)
+        new(args).run
+      end
+
+      def initialize(args)
+        require "pathname"
+
+        @opts         = DEFAULT_OPTS.dup
+        @output       = STDOUT
+        @cmd_root     = Pathname.new Dir.pwd
+        @miqperf_root = Pathname.new File.expand_path("../..", __dir__)
+
+        option_parser.parse!(args)
+        @profile_target  = args.first
+      end
+
+      def run
+        write_tmpfile
+        run_wrapper_script    unless @opts[:debug]
+        debug_wrapper_script  if @opts[:debug]
+      ensure
+        @run_wrapper.close!   if @run_wrapper
+      end
+
+      private
+
+      def write_tmpfile
+        require "erb"
+        require "tempfile"
+
+        @run_wrapper = Tempfile.new("run_wrapper")
+        template_path = "../templates/profile_runner_wrapper.rb.erb"
+        template = File.read File.expand_path(template_path, __dir__)
+
+        @run_wrapper.tap do |wrapper|
+          wrapper.write ERB.new(template, nil, "-").result(binding)
+        end
+      end
+
+      def run_wrapper_script
+        @run_wrapper.close # write the contents of the file to the path
+        run_cmd
+      end
+
+      def debug_wrapper_script
+        @run_wrapper.rewind
+        puts "# Outout: #{@output == STDOUT ? "STDOUT" : @opt[:output_file]}"
+        puts "# Wrapper file:  #{@run_wrapper.path}"
+        puts "# CMD:  #{command}"
+        puts "#"
+        puts "# --------------------"
+        puts
+        puts @run_wrapper.read
+      end
+
+      def run_cmd
+        @output = File.open(@opts[:output_file], "w") if @opts[:output_file]
+        env = {"RUBYOPT" => nil} # don't let bundler inject itself without our permission
+        pid = Kernel.spawn env, command, [:out, :err] => output
+        Process.wait pid
+      end
+
+      def command
+        cmd  = "#{Gem.ruby} #{@run_wrapper.path} "
+        cmd += @opts[:passthrough_args] if @opts[:passthrough_args]
+        cmd += @profile_target if @opts[:target_type] == :rake
+        cmd
+      end
+
+      def use_sysproctable?
+        @use_sysproctable ||=
+          begin
+            if @opts[:sysproctable]
+              require "sys/proctable"
+            else
+              false
+            end
+          rescue LoadError
+            false
+          end
+      end
+
+      def using_stackprof?
+        @use_stackprof ||=
+          begin
+            if @opts[:stackprof_enabled]
+              require "stackprof"
+            else
+              false
+            end
+          rescue LoadError
+            false
+          end
+      end
+
+      def miqperf_profile?
+        @opts[:use_miqperf_profile]
+      end
+
+      def option_parser
```

Comment:
> feels like a bunch of these (almost all?)

50% maybe...

The "Target Types" are all specific to this script, the `--require_tree`, `--mem`, and `--time` stuff are all specific to this as well.  A decent amount of the remaining args just setup for the rest of the command (`--args`, `--bundler`, `--output`, `--preload-app`, etc.).  To your, point, some specifics:

### `--stackprof` and options

I think I originally made this with the intent that some of the profilers could run without the extra weight of `ManageIQPeformance`, and `--miq-prof` was probably one of the last things I added (I think I needed to create `ManageIQPerformance.profile` for that to even work.


### `--miq-prof` and options

The extra options around this are for the one off changes for a single run of the script.  Say I didn't want to run with stackprof in this capture, so run it with `--skip-middleware=stackprof` and you don't need to change your config file in doing so.

### `--mem`

Yeah, this could probably live in `ManageIQPerformance::Middleware` as well, just haven't gotten around to it.  Will probably add it in, but it was nice just having it as a one off drop in script as well.

### `--require-tree`

This one really doesn't make sense in the middleware, unlike a lot of the others.

Reply:
+1 to "one off"